Make it possible to build apps for Android versions still in preview

bartekpacia opened the pull request:
Fixes #104170

It makes it possible to use non-integer `compileSdkVersion` in app-level `build.gradle` file.


## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

bartekpacia committed:
make app's android build work on Tiramisu

with:
- compileSdkVersion "android-Tiramisu"
- targetSdkVersion 33

bartekpacia committed:
create resolveSdkVersion()

bartekpacia committed:
remove printlns

bernaferrari commented on packages/flutter_tools/gradle/flutter.gradle:
```diff
@@ -455,10 +455,31 @@ class FlutterPlugin implements Plugin<Project> {
+     * Handles sdkVersion being integers and strings (codenames) (when the platform stability is not yet declared).
+     * Must be updated when new android SDK level is released.
+     */
+    private int resolveSdkVersion(String sdkVersion) {
+        int resolvedSdkVersion;
+        if (sdkVersion.startsWith("android-")) {
+            sdkVersion = sdkVersion.substring(8)
+        }
+
+        if (sdkVersion.isInteger()) {
+            resolvedSdkVersion = sdkVersion.toInteger()
+        } else if (sdkVersion.equals("Tiramisu")) {
+            resolvedSdkVersion = 33
+        }
```
do you want to hard-code it? Is there a better way?

bartekpacia replied:
I'm afraid there isn't. But this would have to be done only when Google releases a new Android preview, so once a year.

Still, it's simply a draft.

blasten replied:
cc @GaryQian - Tiramisu is still a preview, and while it's reasonable to assume Tiramisu=33, there's no guarantee that it will actually be 33.  Since strings are only used to refer to preview versions, it's fine to detect if the type is a string and exit.

bartekpacia replied:
Sure, that makes sense, but it prevents developers from testing their apps on preview versions of Android. Maybe it's possible to add some escape hatch to make it possible to build Flutter apps for preview Android versions?

blasten replied:
yes - detectLowCompileSdkVersionOrNdkVersion can exit in line 461 if `project.android.compileSdkVersion` isn't a int.

bartekpacia replied:
> `detectLowCompileSdkVersionOrNdkVersion` can exit in line 461

by "exit" you mean just assume that string is fine and not do anything, right?

bartekpacia committed:
revert all changes

bartekpacia committed:
implement suggested solution

If compileSdkVersion isn't an int, then detectLowCompileSdkVersionOrNdkVersion just returns

bartekpacia commented:
@blasten Should I get a test exemption from Hixie or should I write tests for it?
